Editorial review: Document the polygon() basic shape round keyword (#45011)
mdn/content: files/en-us/web/css/reference/values/basic-shape/polygon/index.md

We could define the coordinates of a triangle shape like this:

| axis | point 1 | point 2 | point 3 |
| ---- | ------- | ------- | ------- |
| x    | 0%      | 100%    | 100%    |
| y    | 0%      | 0%      | 100%    |

We can apply those coordinates to the CSS {{cssxref("clip-path")}} property in a `polygon()` function as follows:

```css
clip-path: polygon(0% 0%, 100% 0%, 100% 100%);
```

This creates a triangle shape that covers half the area of its parent container by specifying the coordinates of three of its four corners: top-left (`0% 0%`), top-right (`100% 0%`), and bottom-right (`100% 100%`). If we assume a 200x200px container with a green background:

```html hidden live-sample___basic
<div class="box"></div>
```

```css hidden live-sample___basic
.box {
  width: 200px;
  height: 200px;
  background-color: green;
  clip-path: polygon(0% 0%, 100% 0%, 100% 100%);
}
```

{{EmbedLiveSample("basic", '100%', 200)}}

### The effect of point order

> The order in which you define the points matters and can result in different shapes. For example, the following two `clip-path` declarations both specify a `polygon()` function with four points equal to the four corners of the container.

Reviewer: ```suggestion
The order in which you define the points can result in different shapes. For example, the following two `clip-path` declarations both use a `polygon()` function with X/Y coordinate pairs for the container's four corners, but in a different order.
```

```css
clip-path: polygon(0% 0%, 100% 0%, 100% 100%, 0% 100%);
clip-path: polygon(0% 0%, 100% 0%, 0% 100%, 100% 100%);
```

```html hidden live-sample___different-order
<div class="box"></div>
<div class="box box2"></div>
```

```css hidden live-sample___different-order
body {
  display: flex;
  gap: 20px;
}

.box {
  width: 200px;
  height: 200px;
  background-color: purple;
  clip-path: polygon(0% 0%, 100% 0%, 100% 100%, 0% 100%);
}

.box2 {
  clip-path: polygon(0% 0%, 100% 0%, 0% 100%, 100% 100%);
}
```

> The second `polygon()` function has its points specified in a different order from the first. The first function draws a square, while the second draws an hourglass shape.

Reviewer: If the previous suggestion is accepted:
```suggestion
The first function draws a square, while the second draws an hourglass shape.
```

## Syntax

```css-nolint
/* Coordinate list */
polygon(50% 2.4%, 34.5% 33.8%, 0% 38.8%, 25% 63.1%, 19.1% 97.6%)
polygon(0px 0px, 200px 100px, 0px 200px)
polygon(0% 0px, 100% 100px, 0px 100%)
polygon(0 0, 50% 1rem, 100% 2vw, calc(100% - 20px) 100%, 0 100%)

> /* Coordinate list plus fill rule and/or round value */

Reviewer: ```suggestion
/* Coordinate list with fill-rule and/or round value */
```

polygon(nonzero, 0% 0%, 50% 50%, 0% 100%)
polygon(round 20px, 0% 0%, 50% 50%, 0% 100%)
polygon(evenodd round 2em, 0% 0%, 50% 50%, 0% 100%)
```

### Parameters

> The `polygon` function accepts an optional first parameter containing values that modify the polygon's behavior — a {{SVGAttr("fill-rule")}} keyword and/or the `round` keyword followed by a {{cssxref("length")}} value, separated by spaces — plus one or more parameters containing x/y coordinate pairs of {{cssxref("length-percentage")}} values.

Reviewer: ```suggestion
The `polygon()` function accepts an optional first parameter containing values that modify the polygon's appearance — a {{SVGAttr("fill-rule")}} keyword , the `round` keyword followed by a {{cssxref("length")}} value, or both. The components of the first parameter are separated by spaces. The other parameter  is a comma-separated list of space-separated x/y coordinate pairs of {{cssxref("length-percentage")}} values.
```

- [`<fill-rule>`](/en-US/docs/Web/SVG/Reference/Attribute/fill-rule) {{optional_inline}}
  - : A keyword equal to `nonzero` (the default) or `evenodd`, which specifies the algorithm used to fill the polygon shape.
- `round <length>` {{optional_inline}}
  - : The `round` keyword specifies that the polygon should have rounded corners, and the accompanying {{cssxref("length")}} value specifies the radius of those corners.
- {{cssxref("length-percentage")}}

>   - : Each vertex, or point, of the polygon is represented by a pair of `<length-percentage>` values, which give the x/y coordinates of the vertex relative to the shape's [reference box](/en-US/docs/Web/CSS/Guides/Shapes/Using_shape-outside#the_reference_box).

Reviewer: ```suggestion
  - : Each vertex, or point, of the polygon is represented by a space-separated pair of `<length-percentage>` values defining the x/y coordinates of the vertex relative to the shape's [reference box](/en-US/docs/Web/CSS/Guides/Shapes/Using_shape-outside#the_reference_box).
```

> [!NOTE]

> > The SVG [`<polygon>`](/en-US/docs/Web/SVG/Reference/Element/polygon) element has separate attributes for [`fill-rule`](/en-US/docs/Web/SVG/Reference/Attribute/fill-rule) and [`points`](/en-US/docs/Web/SVG/Reference/Attribute/points), and `points` is flexible about the use of space and comma separators. CSS `polygon()` functions require all parameters to be separated by spaces.

Reviewer: this confuses things. I think we should remove it. But if keps, we have to pedantically include which parts are space separated and which are comma-separated.

Author: Oh gosh, yes, that note is horrible, and I'm more than happy to remove it. Gone in the latest commit ;-)

### Return value

A {{cssxref("basic-shape")}} value.

## Description

You can create almost any shape with the `polygon()` function by specifying the x/y coordinates of its vertices, or points, as comma-separated pairs of {{cssxref("length-percentage")}} values:

<code>polygon(x<sub>1</sub> y<sub>1</sub>, x<sub>2</sub> y<sub>2</sub>, x<sub>3</sub> y<sub>3</sub>, x<sub>4</sub> y<sub>4</sub>, x<sub>n</sub> y<sub>n</sub>)</code>

> Although only a single point is required to create a valid `polygon()` function value, at least 3 points are required to create a shape (a triangle). There is no upper limit on the number of points that can be specified. The shape is drawn through the specified points in the order they appear in the function, with a final line being automatically drawn between the last and first points to close up the shape.

Reviewer: ```suggestion
Although only a single point is required to create a valid `polygon()` function value, at least 3 points are required to create a shape (a triangle). There is no upper limit on the number of points that can be specified. The shape is drawn through the specified points in the order they appear in the function, with a final line automatically being drawn between the last and first points to close the shape.
```

```html hidden live-sample___basic
<div class="box"></div>
```

```css hidden live-sample___basic
.box {
  width: 200px;
  height: 200px;
  background-color: green;
  clip-path: polygon(0% 0%, 100% 0%, 100% 100%);
}
```

{{EmbedLiveSample("basic", '100%', 200)}}

### The effect of point order

The order in which you define the points matters and can result in different shapes. For example, the following two `clip-path` declarations both specify a `polygon()` function with four points equal to the four corners of the container.

```css
clip-path: polygon(0% 0%, 100% 0%, 100% 100%, 0% 100%);
clip-path: polygon(0% 0%, 100% 0%, 0% 100%, 100% 100%);
```

```html hidden live-sample___different-order
<div class="box"></div>
<div class="box box2"></div>
```

```css hidden live-sample___different-order
body {
  display: flex;
  gap: 20px;
}

.box {
  width: 200px;
  height: 200px;
  background-color: purple;
  clip-path: polygon(0% 0%, 100% 0%, 100% 100%, 0% 100%);
}

.box2 {
  clip-path: polygon(0% 0%, 100% 0%, 0% 100%, 100% 100%);
}
```

> However, the second `polygon()` function has its points specified in a different order to the first. The first function draws a square, whereas the second draws an hourglass shape.

Reviewer: ```suggestion
The second `polygon()` function has its points specified in a different order from the first. The first function draws a square, while the second draws an hourglass shape.
```

{{EmbedLiveSample("different-order", '100%', 200)}}

### Specifying polygon modifiers

> The `polygon()` function can accept an optional first parameter containing values that modify the rendering of the resulting shape. The parameter value can include any or all of the following, separated by spaces:

Reviewer: ```suggestion
The `polygon()` function accepts an optional first parameter modifying the rendering of the resulting shape. The parameter value can include either or both of the following, separated by spaces:
```

> - A [`<fill-rule>`](/en-US/docs/Web/SVG/Reference/Attribute/fill-rule) keyword equal to `nonzero` (the default) or `evenodd`, which specifies the algorithm used to fill the polygon shape. This rarely has much of an effect.

Reviewer: ```suggestion
- A [`<fill-rule>`](/en-US/docs/Web/SVG/Reference/Attribute/fill-rule) keyword equal to `nonzero` (the default) or `evenodd`, which specifies the algorithm used to fill the polygon shape. This only has an effect when the lines drawn between coordinate values overlap.
```

- The `round` keyword followed by a {{cssxref("length")}} value. This specifies that the polygon should have rounded corners, with the `<length>` value specifying the radius of those corners.

> For example, we could modify the earlier triangle example to use the following `clip-path` declaration:

Reviewer: ```suggestion
For example, we could expand upon the previous triangle example and add rounded corners:
```

```css
clip-path: polygon(round 20px, 0% 0%, 100% 0%, 100% 100%);
```

```html hidden live-sample___basic-rounded
<div class="box"></div>
```

```css hidden live-sample___basic-rounded
.box {
  width: 200px;
  height: 200px;
  background-color: green;
  clip-path: polygon(round 20px, 0% 0%, 100% 0%, 100% 100%);
}
```

This results in the same triangle shape, but with `20px` radius rounded corners:

{{EmbedLiveSample("basic-rounded", '100%', 200)}}

> [!NOTE]

> > In each case, the maximum corner radius is clamped to the smaller of `tan(corner-angle/2) * (segment-length / 2)` evaluated against both line segments forming the corner. This ensures that the applied corner radius is never larger than half of any line segment length. [The specification](https://drafts.csswg.org/css-shapes-1/#funcdef-basic-shape-polygon) contains further details for those interested.

Reviewer: ```suggestion
> In each case, the corner radius is clamped to ensure it is never larger than half of any line segment length. The maximum corner radius is clamped to the smaller of `tan(corner-angle/2) * (segment-length / 2)` evaluated against both line segments forming the corner. [The specification](https://drafts.csswg.org/css-shapes-1/#funcdef-basic-shape-polygon) contains further details for those interested.
```
putting the grokable part first, and the nerdy part after for those interested (want people to know what it does, and the equation first may intimidate some from reading the rest of the paragraph)

#### Result

{{EmbedLiveSample("triangle", '100%', 400)}}

The coordinates for the triangle are the top-right corner (`100% 0%`), the center point (`50% 50%`), and the bottom-right corner (`100% 100%`) of the container.

### Create a rounded star

In this example, we create a star shape and use the `round` keyword to round its corners.

#### HTML

```html live-sample___star
<div class="star"></div>
```

#### CSS

```css live-sample___star
.star {
  width: 400px;
  height: 400px;
  background-color: hotpink;
  clip-path: polygon(
    round 20px,
    50% 5%,
    60.85% 27.48%,
    85.22% 21.99%,
    74.38% 44.44%,
    93.88% 60.01%,
    69.57% 65.56%,
    69.53% 90.55%,
    50% 75%,
    30.47% 90.55%,
    30.43% 65.56%,
    6.12% 60.01%,
    25.62% 44.44%,
    14.78% 21.99%,
    39.15% 27.48%
  );
}
```

```css hidden live-sample___basic-rounded live-sample___star
@supports not (clip-path: polygon(round 20px, 0% 0%, 100% 0%, 100% 100%)) {
  body::before {
    font-family: sans-serif;

>     content: "Your browser does not support the polygon() round keyword.";

Reviewer: ```suggestion
    content: "Your browser does not support the polygon()'s round keyword.";
```
